From 5bc2330869bd651a7c099110ee38a03c6acd7dfe Mon Sep 17 00:00:00 2001
From: Jes Sorensen <Jes.Sorensen@redhat.com>
Date: Mon, 1 Sep 2014 09:25:41 -0400
Subject: [md] raid5: Fix possible confusion when multiple write errors occur

Message-id: <1409563584-9000-40-git-send-email-Jes.Sorensen@redhat.com>
Patchwork-id: 90006
O-Subject: [RHEL7.1 PATCH 39/82] md/raid5: Fix possible confusion when multiple write errors occur.
Bugzilla: 1085530
RH-Acked-by: Tony Camuso <tcamuso@redhat.com>
RH-Acked-by: Prarit Bhargava <prarit@redhat.com>

From: NeilBrown <neilb@suse.de>

commit 5d8c71f9e5fbdd95650be00294d238e27a363b5c
    md: raid5 crash during degradation

Fixed a crash in an overly simplistic way which could leave
R5_WriteError or R5_MadeGood set in the stripe cache for devices
for which it is no longer relevant.
When those devices are removed and spares added the flags are still
set and can cause incorrect behaviour.

commit 14a75d3e07c784c004b4b44b34af996b8e4ac453
    md/raid5: preferentially read from replacement device if possible.

Fixed the same bug if a more effective way, so we can now revert
the original commit.

Reported-and-tested-by: Alexander Lyakas <alex.bolshoy@gmail.com>
Cc: stable@vger.kernel.org (3.2+ - 3.2 will need a different fix though)
Fixes: 5d8c71f9e5fbdd95650be00294d238e27a363b5c
Signed-off-by: NeilBrown <neilb@suse.de>
(cherry picked from commit 1cc03eb93245e63b0b7a7832165efdc52e25b4e6)
Signed-off-by: Jes Sorensen <Jes.Sorensen@redhat.com>
---

Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/drivers/md/raid5.c b/drivers/md/raid5.c
index ab595b2..ab6ae11 100644
--- a/drivers/md/raid5.c
+++ b/drivers/md/raid5.c
@@ -3608,7 +3608,7 @@ static void analyse_stripe(struct stripe_head *sh, struct stripe_head_state *s)
     */
    set_bit(R5_Insync, &dev->flags);
 
-  if (rdev && test_bit(R5_WriteError, &dev->flags)) {
+  if (test_bit(R5_WriteError, &dev->flags)) {
    /* This flag does not apply to '.replacement'
     * only to .rdev, so make sure to check that*/
    struct md_rdev *rdev2 = rcu_dereference(
@@ -3621,7 +3621,7 @@ static void analyse_stripe(struct stripe_head *sh, struct stripe_head_state *s)
    } else
     clear_bit(R5_WriteError, &dev->flags);
   }
-  if (rdev && test_bit(R5_MadeGood, &dev->flags)) {
+  if (test_bit(R5_MadeGood, &dev->flags)) {
    /* This flag does not apply to '.replacement'
     * only to .rdev, so make sure to check that*/
    struct md_rdev *rdev2 = rcu_dereference(
-- 
1.7.1